src/Sparrow/Allocators.cs

```diff
@@ -0,0 +1,524 @@
+        }
+
+        public void Reset()
+        {
+            _allocator.Reset(ref _allocator);
+        }
+
+        public void Dispose()
+        {
+            if (_disposeFlag.Raise())
+                _allocator.Dispose(ref _allocator);
+
+            GC.SuppressFinalize(this);
+        }
+
+        public void LowMemory()
+        {
+            if (_allocator is ILowMemoryHandler<TAllocator> a)
+                a.NotifyLowMemory(ref _allocator);
+        }
+
+        public void LowMemoryOver()
+        {
+            if (_allocator is ILowMemoryHandler<TAllocator> a)
+                a.NotifyLowMemoryOver(ref _allocator);
+        }
+    }
+
+    public sealed class FixedSizeAllocator<TAllocator> : IDisposable, ILowMemoryHandler
+        where TAllocator : struct, IAllocator<TAllocator, Pointer>, IAllocator
+    {
+        private int _blockSize;
+        private TAllocator _allocator;
+        private readonly SingleUseFlag _disposeFlag = new SingleUseFlag();
+
+        ~FixedSizeAllocator()
+        {
+            if (_allocator is ILifecycleHandler<TAllocator> a)
+                a.BeforeFinalization(ref _allocator);
+
+            Dispose();
+        }
+
+        public void Initialize<TConfig>(TConfig options)
+            where TConfig : struct, IAllocatorOptions
+        {
+
+            if (!typeof(IFixedSizeAllocatorOptions).IsAssignableFrom(typeof(TConfig)))
+                throw new NotSupportedException($"{nameof(TConfig)} is not compatible with {nameof(TConfig)}");
+
+            this._blockSize = ((IFixedSizeAllocatorOptions)options).BlockSize;
+
+            if (_allocator is ILifecycleHandler<TAllocator> a)
+                a.BeforeInitialize(ref _allocator);
+
+            _allocator.Initialize(ref _allocator);
+            _allocator.Configure(ref _allocator, ref options);
+
+            if (_allocator is ILowMemoryHandler<TAllocator>)
+                LowMemoryNotification.Instance.RegisterLowMemoryHandler(this);
+
+            if (_allocator is ILifecycleHandler<TAllocator> b)
+                b.AfterInitialize(ref _allocator);
+        }
+
+        public long Allocated
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get { return _allocator.TotalAllocated; }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Pointer Allocate()
+        {
+            unsafe
+            {
+                var ptr = _allocator.Allocate(ref _allocator, this._blockSize);
+                if (_allocator is ILifecycleHandler<TAllocator> a)
+                    a.BeforeInitialize(ref _allocator);
+
+                return ptr;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Pointer<TType> Allocate<TType>() where TType : struct
+        {
+            unsafe
+            {
+                var ptr = _allocator.Allocate(ref _allocator, this._blockSize * Unsafe.SizeOf<TType>());
```

Comment:
How is multi threading handled in this case?

Reply:
Underlying allocator must support it. The shell is Thread Safe.

Reply:
Currently it is like that, I modeled it based on current operation. We can implement underlying thread safe and also thread sharing allocators which would fixed that if we want. The idea of composability is that you can implement behaviors based on the requirements and pay the cost only for those cases. Now the question is: do we need multithread safe allocators for pools and fragments? If that is so, we can implement those too

Reply:
I'm perfectly fine with stating that allocators being single threaded, but it needs to be stated clearly as such.

Reply:
In particular, what allocators _are_ safe for multi threading and can be re-used.

Reply:
Only multi-threaded safe allocators now are:
- NativeAllocator
- ThreadAffinePoolAllocator